pkg/cmd/roachtest/cluster.go

```diff
@@ -973,6 +966,9 @@ func (f *clusterFactory) newCluster(
 			// or a destroy from the previous iteration failed.
 			return nil, nil, err
 		}
+		if errors.HasType(err, (*roachprod.MalformedClusterNameError)(nil)) {
```

Comment:
Let's add a unit test to cover this case.

Reply:
I gave it a shot, but I wasn't sure how to feasibly do this without mocking a ton of things. There already is a way to mock cluster creation via `TestSpec` + `r.MakeClusterSpec(0)`, but that skips `roachprod.Create` and this entire retry loop where we validate the cluster name.

I could mock `roachprod.Create` to just return a `MalformedclusterNameError`, but I don't think that test would be very useful.

Reply:
> I could mock roachprod.Create to just return a MalformedclusterNameError, but I don't think that test would be very useful.

That would have been my suggestion as well. I think it would be useful to unit test `roachprod.Create` especially in the context of [1]; i.e., `ClusterAlreadyExistsError`, `MalformedClusterNameError` and potentially future type of errors are non-retryable, so we should be able to cover the fact via a unit test. For now, the mock could randomly return the two non-retryable errors, and we assert that the mock is invoked exactly once, i.e., no retries. Wdyt?

[1] https://github.com/cockroachdb/cockroach/issues/114523

Reply:
Done. I was trying to add a case where it _didn't_ error out and that was quite tricky to do without mocking a bunch of things. Mocking the two non-retryable errors cases was simple though.